sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnInvokers.java

```diff
@@ -0,0 +1,460 @@
+      Class<? extends DoFn> fnClass) {
+    Constructor<?> constructor = BYTE_BUDDY_INVOKER_CONSTRUCTOR_CACHE.get(fnClass);
+    if (constructor != null) {
+      return constructor;
+    }
+    Class<? extends DoFnInvoker<?, ?>> wrapperClass = createWrapperClass(fnClass);
+    try {
+      constructor = wrapperClass.getConstructor(fnClass);
+    } catch (IllegalArgumentException | NoSuchMethodException | SecurityException e) {
+      throw new RuntimeException(e);
+    }
+    BYTE_BUDDY_INVOKER_CONSTRUCTOR_CACHE.put(fnClass, constructor);
+    return constructor;
+  }
+
+  private static Class<? extends DoFnInvoker<?, ?>> createWrapperClass(
+      Class<? extends DoFn> clazz) {
+    final TypeDescription clazzDescription = new TypeDescription.ForLoadedType(clazz);
+
+    DoFnSignature signature = DoFnReflector.getSignature(clazz);
+
+    DynamicType.Builder<?> builder =
+        new ByteBuddy()
+            // Create subclasses inside the target class, to have access to
+            // private and package-private bits
+            .with(
+                new NamingStrategy.SuffixingRandom("auxiliary") {
+                  @Override
+                  public String subclass(TypeDescription.Generic superClass) {
+                    return super.name(clazzDescription);
+                  }
+                })
+            // Create a subclass of DoFnInvoker
+            .subclass(DoFnInvoker.class, ConstructorStrategy.Default.NO_CONSTRUCTORS)
+            .defineField(
+                FN_DELEGATE_FIELD_NAME, clazz, Visibility.PRIVATE, FieldManifestation.FINAL)
+            .defineConstructor(Visibility.PUBLIC)
+            .withParameter(clazz)
+            .intercept(new InvokerConstructor())
+            // Delegate processElement(), startBundle() and finishBundle() to the fn.
+            .method(ElementMatchers.named("invokeProcessElement"))
+            .intercept(new ProcessElementDelegation(signature.getProcessElement()))
+            .method(ElementMatchers.named("invokeStartBundle"))
+            .intercept(
+                signature.getStartBundle() == null
+                    ? new NoopMethodImplementation()
+                    : new BundleMethodDelegation(signature.getStartBundle()))
+            .method(ElementMatchers.named("invokeFinishBundle"))
+            .intercept(
+                signature.getFinishBundle() == null
+                    ? new NoopMethodImplementation()
+                    : new BundleMethodDelegation(signature.getFinishBundle()));
+
+    DynamicType.Unloaded<?> unloaded = builder.make();
+
+    @SuppressWarnings("unchecked")
+    Class<? extends DoFnInvoker<?, ?>> res =
+        (Class<? extends DoFnInvoker<?, ?>>)
+            unloaded
+                .load(DoFnInvokers.class.getClassLoader(), ClassLoadingStrategy.Default.INJECTION)
+                .getLoaded();
+    return res;
+  }
+
+  private static class NoopMethodImplementation implements Implementation {
```

Comment:
`/** Implement an invoker method by doing nothing and immediately returning void. */`